File: packages/twilio_flutter_conversations/ios/Classes/Bridge/ConversationsEventMapper.swift
import Foundation
import TwilioConversationsClient

/// Maps Twilio Conversations SDK models to pigeon DTOs.
final class ConversationsEventMapper {
  func mapSynchronizationStatus(
    _ status: TCHClientSynchronizationStatus
  ) -> ClientSynchronizationStatus {
    switch status {
    case .started:
      return .started
    case .conversationsListCompleted:
      return .conversationsListCompleted
    case .completed:
      return .completed
    case .failed:
      return .failed
    default:
      return .unknown
    }
  }

  func mapConnectionState(_ state: TCHClientConnectionState) -> ClientConnectionState {
    switch state {
    case .connecting:
      return .connecting
    case .connected:
      return .connected
    case .disconnected:
      return .disconnected
    case .denied:
      return .denied
    case .error:
      return .error
    @unknown default:
      if String(describing: state).lowercased().contains("fatal") {
        return .fatal
      }
      return .unknown
    }
  }

  func mapConversationSynchronizationStatus(
    _ status: TCHConversationSynchronizationStatus
  ) -> ConversationSynchronizationStatus {
    switch status {
    case .none:
      return .none
    case .identifier:
      return .identifier
    case .metadata:
      return .metadata
    case .all:
      return .all
    case .failed:
      return .failed
    default:
      return .unknown
    }
  }

  func mapConversation(_ conversation: TCHConversation) -> ConversationDto {
    return ConversationDto(
      sid: conversation.sid ?? "",
      uniqueName: conversation.uniqueName ?? "",
      friendlyName: conversation.friendlyName ?? "",
      lastMessageIndex: conversation.lastMessageIndex?.intValue
    )
  }

  func mapMessage(_ message: TCHMessage, conversationSid: String) -> MessageDto {
    let base = MessageDto(
      sid: message.sid ?? "",
      conversationSid: conversationSid,
      author: message.author ?? "",
      body: message.body ?? "",
      messageIndex: Int64(message.index),
      dateCreatedEpochMs: Int64((message.dateCreatedAsDate?.timeIntervalSince1970 ?? 0) * 1000),
      contentType: .text,
      mediaAttachments: nil,
      attributesJson: MessageAttributesJson.json(from: message)
    )
    return MessageMediaMapping.enrichMessageDto(base: base, message: message)
  }

  func mapParticipant(_ participant: TCHParticipant, conversationSid: String) -> ParticipantDto {
    return ParticipantDto(
      sid: participant.sid ?? "",
      identity: participant.identity ?? "",
      conversationSid: conversationSid
    )
  }

  func mapUser(_ user: TCHUser) -> UserDto {
    return UserDto(
      identity: user.identity ?? "",
      friendlyName: user.friendlyName
    )
  }

  func synchronizationEvent(
    status: TCHClientSynchronizationStatus
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .clientSynchronizationStatusUpdated,
      synchronizationStatus: mapSynchronizationStatus(status),
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func connectionStateEvent(state: TCHClientConnectionState) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .connectionStateChanged,
      synchronizationStatus: nil,
      connectionState: mapConnectionState(state),
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func conversationAddedEvent(_ conversation: TCHConversation) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .conversationAdded,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: mapConversation(conversation),
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func conversationUpdatedEvent(
    _ conversation: TCHConversation,
    updated: TCHConversationUpdate
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .conversationUpdated,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: mapConversation(conversation),
      message: nil,
      participant: nil,
      user: nil,
      updateReason: String(describing: updated),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func conversationDeletedEvent(_ conversation: TCHConversation) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .conversationDeleted,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: mapConversation(conversation),
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func conversationSynchronizationEvent(
    _ conversation: TCHConversation
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .conversationSynchronizationUpdated,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: mapConversationSynchronizationStatus(
        conversation.synchronizationStatus
      ),
      conversation: mapConversation(conversation),
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func messageAddedEvent(
    _ message: TCHMessage,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .messageAdded,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: mapMessage(message, conversationSid: conversationSid),
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func messageUpdatedEvent(
    _ message: TCHMessage,
    conversationSid: String,
    updated: TCHMessageUpdate
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .messageUpdated,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: mapMessage(message, conversationSid: conversationSid),
      participant: nil,
      user: nil,
      updateReason: String(describing: updated),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func messageDeletedEvent(
    _ message: TCHMessage,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .messageDeleted,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: mapMessage(message, conversationSid: conversationSid),
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func participantAddedEvent(
    _ participant: TCHParticipant,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .participantAdded,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: mapParticipant(participant, conversationSid: conversationSid),
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func participantUpdatedEvent(
    _ participant: TCHParticipant,
    conversationSid: String,
    updated: TCHParticipantUpdate
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .participantUpdated,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: mapParticipant(participant, conversationSid: conversationSid),
      user: nil,
      updateReason: String(describing: updated),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func participantDeletedEvent(
    _ participant: TCHParticipant,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .participantDeleted,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: mapParticipant(participant, conversationSid: conversationSid),
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func typingStartedEvent(
    _ participant: TCHParticipant,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .typingStarted,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: mapParticipant(participant, conversationSid: conversationSid),
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func typingEndedEvent(
    _ participant: TCHParticipant,
    conversationSid: String
  ) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .typingEnded,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: mapParticipant(participant, conversationSid: conversationSid),
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func userUpdatedEvent(_ user: TCHUser, updated: TCHUserUpdate) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .userUpdated,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: mapUser(user),
      updateReason: String(describing: updated),
      errorCode: nil,
      errorMessage: nil
    )
  }

  func userSubscribedEvent(_ user: TCHUser) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .userSubscribed,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: mapUser(user),
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func userUnsubscribedEvent(_ user: TCHUser) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .userUnsubscribed,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: mapUser(user),
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func tokenAboutToExpireEvent() -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .tokenAboutToExpire,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func tokenExpiredEvent() -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .tokenExpired,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func notificationSubscribedEvent() -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .notificationSubscribed,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: nil,
      errorMessage: nil
    )
  }

  func errorEvent(code: String, message: String) -> ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type: .error,
      synchronizationStatus: nil,
      connectionState: nil,
      conversationSyncStatus: nil,
      conversation: nil,
      message: nil,
      participant: nil,
      user: nil,
      updateReason: nil,
      errorCode: code,
      errorMessage: message
    )
  }
}

File: packages/twilio_flutter_conversations/android/src/main/kotlin/com/twilioflutter/conversations/bridge/ConversationsEventMapper.kt
package com.twilioflutter.conversations.bridge

import com.twilio.conversations.Conversation
import com.twilio.conversations.ConversationsClient
import com.twilio.conversations.Message
import com.twilio.conversations.Participant
import com.twilio.conversations.User
import com.twilioflutter.conversations.pigeon.ClientConnectionState
import com.twilioflutter.conversations.pigeon.ClientSynchronizationStatus
import com.twilioflutter.conversations.pigeon.ConversationDto
import com.twilioflutter.conversations.pigeon.ConversationSynchronizationStatus
import com.twilioflutter.conversations.pigeon.ConversationsEventType
import com.twilioflutter.conversations.pigeon.ConversationsNativeEvent
import com.twilioflutter.conversations.pigeon.MessageContentType
import com.twilioflutter.conversations.pigeon.MessageDto
import com.twilioflutter.conversations.pigeon.ParticipantDto
import com.twilioflutter.conversations.pigeon.UserDto

/** Maps Twilio Conversations SDK models to pigeon DTOs. */
class ConversationsEventMapper {
  fun mapSynchronizationStatus(
    status: ConversationsClient.SynchronizationStatus,
  ): ClientSynchronizationStatus {
    return when (status) {
      ConversationsClient.SynchronizationStatus.STARTED ->
        ClientSynchronizationStatus.STARTED
      ConversationsClient.SynchronizationStatus.CONVERSATIONS_COMPLETED ->
        ClientSynchronizationStatus.CONVERSATIONS_LIST_COMPLETED
      ConversationsClient.SynchronizationStatus.COMPLETED ->
        ClientSynchronizationStatus.COMPLETED
      ConversationsClient.SynchronizationStatus.FAILED ->
        ClientSynchronizationStatus.FAILED
      ConversationsClient.SynchronizationStatus.SYNCHRONIZATION_DISABLED ->
        ClientSynchronizationStatus.UNKNOWN
      else -> ClientSynchronizationStatus.UNKNOWN
    }
  }

  fun mapConnectionState(
    state: ConversationsClient.ConnectionState,
  ): ClientConnectionState {
    return when (state) {
      ConversationsClient.ConnectionState.CONNECTING ->
        ClientConnectionState.CONNECTING
      ConversationsClient.ConnectionState.CONNECTED ->
        ClientConnectionState.CONNECTED
      ConversationsClient.ConnectionState.DISCONNECTED ->
        ClientConnectionState.DISCONNECTED
      ConversationsClient.ConnectionState.DENIED ->
        ClientConnectionState.DENIED
      ConversationsClient.ConnectionState.ERROR ->
        ClientConnectionState.ERROR
      ConversationsClient.ConnectionState.FATAL_ERROR ->
        ClientConnectionState.FATAL
      else -> ClientConnectionState.UNKNOWN
    }
  }

  fun mapConversationSynchronizationStatus(
    conversation: Conversation,
  ): ConversationSynchronizationStatus {
    return when (conversation.synchronizationStatus) {
      Conversation.SynchronizationStatus.NONE ->
        ConversationSynchronizationStatus.NONE
      Conversation.SynchronizationStatus.IDENTIFIER ->
        ConversationSynchronizationStatus.IDENTIFIER
      Conversation.SynchronizationStatus.METADATA ->
        ConversationSynchronizationStatus.METADATA
      Conversation.SynchronizationStatus.ALL ->
        ConversationSynchronizationStatus.ALL
      Conversation.SynchronizationStatus.FAILED ->
        ConversationSynchronizationStatus.FAILED
      else -> ConversationSynchronizationStatus.UNKNOWN
    }
  }

  fun mapConversation(conversation: Conversation): ConversationDto {
    return ConversationDto(
      sid = conversation.sid,
      uniqueName = conversation.uniqueName ?: "",
      friendlyName = conversation.friendlyName ?: "",
      lastMessageIndex = conversation.lastMessageIndex,
    )
  }

  fun mapMessage(message: Message, conversationSid: String): MessageDto {
    val base =
      MessageDto(
        sid = message.sid,
        conversationSid = conversationSid,
        author = message.author ?: "",
        body = message.body ?: "",
        messageIndex = message.messageIndex,
        dateCreatedEpochMs = message.dateCreatedAsDate?.time ?: 0L,
        contentType = MessageContentType.TEXT,
        attributesJson = MessageAttributesJson.jsonFromMessage(message),
      )
    return MessageMediaMapping.enrichMessageDto(base, message)
  }

  fun mapParticipant(participant: Participant, conversationSid: String): ParticipantDto {
    return ParticipantDto(
      sid = participant.sid,
      identity = participant.identity ?: "",
      conversationSid = conversationSid,
    )
  }

  fun mapUser(user: User): UserDto {
    return UserDto(
      identity = user.identity ?: "",
      friendlyName = user.friendlyName,
    )
  }

  fun synchronizationEvent(
    status: ConversationsClient.SynchronizationStatus,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CLIENT_SYNCHRONIZATION_STATUS_UPDATED,
      synchronizationStatus = mapSynchronizationStatus(status),
    )
  }

  fun connectionStateEvent(
    state: ConversationsClient.ConnectionState,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CONNECTION_STATE_CHANGED,
      connectionState = mapConnectionState(state),
    )
  }

  fun conversationAddedEvent(conversation: Conversation): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CONVERSATION_ADDED,
      conversation = mapConversation(conversation),
    )
  }

  fun conversationUpdatedEvent(
    conversation: Conversation,
    reason: Conversation.UpdateReason,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CONVERSATION_UPDATED,
      conversation = mapConversation(conversation),
      updateReason = reason.name,
    )
  }

  fun conversationDeletedEvent(conversation: Conversation): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CONVERSATION_DELETED,
      conversation = mapConversation(conversation),
    )
  }

  fun conversationSynchronizationEvent(
    conversation: Conversation,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.CONVERSATION_SYNCHRONIZATION_UPDATED,
      conversation = mapConversation(conversation),
      conversationSyncStatus = mapConversationSynchronizationStatus(conversation),
    )
  }

  fun messageAddedEvent(message: Message, conversationSid: String): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.MESSAGE_ADDED,
      message = mapMessage(message, conversationSid),
    )
  }

  fun messageUpdatedEvent(
    message: Message,
    conversationSid: String,
    reason: Message.UpdateReason,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.MESSAGE_UPDATED,
      message = mapMessage(message, conversationSid),
      updateReason = reason.name,
    )
  }

  fun messageDeletedEvent(
    message: Message,
    conversationSid: String,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.MESSAGE_DELETED,
      message = mapMessage(message, conversationSid),
    )
  }

  fun participantAddedEvent(
    participant: Participant,
    conversationSid: String,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.PARTICIPANT_ADDED,
      participant = mapParticipant(participant, conversationSid),
    )
  }

  fun participantUpdatedEvent(
    participant: Participant,
    conversationSid: String,
    reason: Participant.UpdateReason,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.PARTICIPANT_UPDATED,
      participant = mapParticipant(participant, conversationSid),
      updateReason = reason.name,
    )
  }

  fun participantDeletedEvent(
    participant: Participant,
    conversationSid: String,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.PARTICIPANT_DELETED,
      participant = mapParticipant(participant, conversationSid),
    )
  }

  fun typingStartedEvent(
    participant: Participant,
    conversationSid: String,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.TYPING_STARTED,
      participant = mapParticipant(participant, conversationSid),
    )
  }

  fun typingEndedEvent(
    participant: Participant,
    conversationSid: String,
  ): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.TYPING_ENDED,
      participant = mapParticipant(participant, conversationSid),
    )
  }

  fun userUpdatedEvent(user: User, reason: User.UpdateReason): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.USER_UPDATED,
      user = mapUser(user),
      updateReason = reason.name,
    )
  }

  fun userSubscribedEvent(user: User): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.USER_SUBSCRIBED,
      user = mapUser(user),
    )
  }

  fun userUnsubscribedEvent(user: User): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.USER_UNSUBSCRIBED,
      user = mapUser(user),
    )
  }

  fun tokenAboutToExpireEvent(): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.TOKEN_ABOUT_TO_EXPIRE,
    )
  }

  fun tokenExpiredEvent(): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.TOKEN_EXPIRED,
    )
  }

  fun notificationSubscribedEvent(): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.NOTIFICATION_SUBSCRIBED,
    )
  }

  fun errorEvent(code: String, message: String): ConversationsNativeEvent {
    return ConversationsNativeEvent(
      type = ConversationsEventType.ERROR,
      errorCode = code,
      errorMessage = message,
    )
  }
}
